### ISSUE TITLE
git-bindir is being ignored on Windows

### ISSUE THREAD

**Issue by zilti** [git-cola/git-cola#1258]
Since I currently have the misfortune of having to use Windows (and at that one without admin rights) I installed both Git and Git-Cola there. Unfortunately Git doesn't stay on the path, so as advised in the error dialog box, I tried creating the git-bindir file with the following content:
```
C:\Users\ziltenerd\AppData\Local\Programs\Git\bin
```

I created it both at the `C:\Users\ziltenerd\.config\git-cola` directory that gets auto-created as well as at `C:\Users\ziltenerd\AppData\Roaming\.config\git-cola`. Both had no effect at all.

**Comment by davvid** [comment]
This has been a curiosity for me, but I'll need your help diagnosing this. I think this is basically a duplicate of #927 .

If you could kindly hop over there and diagnose what `os.path.expanduser('~/.config/git-cola/git-bindir')` expands to in your environment then we might be able to solve it.

The relevant code is in `cola/app.py` (search for `git-bindir`) and we're just asking python to expand the value.

In case you just need something working, I just pushed a fix that makes cola honor `XDG_CONFIG_HOME` in your environment. That'll control the directory where it expects to find `git-bindir` file, in case that's the root cause.

Thanks for the heads-up. If you can help with testing in #927 then I'd be open to resurrecting the old patch that I whipped up in that issue, but I don't want to bifurcate the threads so I'm closing this issue.

There wasn't anyone available to test it, so I wasn't comfortable applying the patch that I worked on back then, but if you're able to run from git and test stuff out then we can probably get this sorted pretty easily.

**Comment by zilti** [comment]
I ran it:

```python
>>> import os
>>> os.path.expanduser('~/.config/git-cola/git-bindir')
'C:\\Users\\ziltenerd/.config/git-cola/git-bindir'
```

So it is in the same directory as the `settings` file.

**Comment by zilti** [comment]
I've tried setting XDG_CONFIG_HOME to `C:\Users\ziltenerd\.config` but that made no difference.

**Comment by davvid** [comment]
Quick question: are you launching cola from the shell or from the shortcut icon?

I'm reading through the code and our `main()` function doesn't do the win32 git-bindir $PATH fixups. If you're launching it through the `git-cola` command then I suspect that I need to change where we initialize that.

The launcher icons all go through `app.shortcut_launch()` which is where the $PATH adjustment happens (via `winmain()`).

Let me know if you're launching it a different way. If so we probably just need to adjust where we do that initialization

**Comment by davvid** [comment]
Thanks for those details -- I was hoping that 4726cd5f8885ecc8b351bfcd3f5e1308c0cc62f5 might help because maybe we weren't always adding the configured `git-bindir` to the `$PATH`, but since you're launching from the shortcut icon then it should have already been going through the `$PATH` setup code.

In any case, I've adjusted the logic so that this code path is always active. I kinda doubt that'll make a difference.

The other thing that changed is that the config reader was just rewritten to ensure that we can always read gitconfig-provided configuration values. The fixes in #927 seemed related because they both involve reading from `~` (`$HOME`) but they don't really affect the `git-bindir` code path because that reads from the `$XDG_CONFIG_HOME/git-cola/git-bindir` file directly.

Thus, I've re-opened this issue since what we're hitting here seems unique. This is a bit surprising.

All we're really doing is reading `~/.config/git-cola/git-bindir` and then adding the value we read from the file to the front of the `PATH` environment variable.

All of the points to inspect are in `cola/app.py`. The two functions of interest are `def find_git()` and `def prepend_path(path)`.

`find_git` has an early-out if it doesn't think we're on Windows. What does this print in your install?

```
import sys
print(sys.platform)
```

We only consider `win32` and `cygwin` values in `def is_win32()` from `cola/utils.py` -- is there a chance you have a different value? If so that would cause `find_git()` to return `None` and it won't add the value to the `PATH`. That said, I tried looking at the value in a VM and I saw `win32` there, so probably not that.

The other place it could fail is in `def prepend_path(path)`. This function is fairly straightforward. It reads the `git-bindir` file and prepends to the PATH environment variable. It's kinda surprising that this function would fail, but anything is possible.

-------

Here's a some other details that might help out.

Earlier you mentioned that you could set environment variables (XDG_CONFIG_HOME).

If you can do that, are you able to prepend to `PATH` locally and then see if that works sans `git-bindir`?

Alternatively, I noticed that we do support one other environment variable in `cola/git.py`:

```
GIT = core.getenv('GIT_COLA_GIT', 'git')
```

If you define `GIT_COLA_GIT` in your environment and point it directly to the full, absolute path to `git.exe` then we will use that value when executing git. When undefined we just use `git` and expect to find it in $PATH.

In theory we should be able to get this working by messing with just the `GIT_COLA_GIT` environment variable. There's much less machinery involved in that code path so it might shed some light on what's going on.

Can you try using that environment variable to see if it helps? Note that you will need to launch from a Git Bash shell in order for it to pickup your shell environment.

Are you able to run from the source tree? If so it'd be worth testing the latest from git in case the recent fixes helped in some way, but `GIT_COLA_GIT` has been around for a very long time so if you can control the environment then that should be usable with your current version.

Thanks again for trying this stuff out.
